etl-pipeline/processes/record_file_saver.py

```diff
@@ -26,20 +29,46 @@ def save_record_files(self, record: Record):
 
     def store_file(self, part: Part):
         try:
-            file_contents = self.get_file_contents(part.source_url)
-            self.storage_manager.put_object(file_contents, part.file_key, part.file_bucket)
-            del file_contents
-
-            if '.epub' in part.file_key:
-                file_root = '.'.join(part.file_key.split('.')[:-1])
-
-                web_pub_manifest = self.generate_webpub(file_root)
-                self.storage_manager.put_object(web_pub_manifest, f'{file_root}/manifest.json', self.file_bucket)
-
-            logger.info(f'Stored file {part.file_key} from {part.source_url}')
+            if 'drive.google.com' in part.source_url:
+                self._store_file_from_drive(part)
+            elif part.source_file_key and part.source_file_bucket:
+                self._copy_file(part)
+            else:
+                file_contents = self.get_file_contents(part.source_url)
+                self.storage_manager.put_object(file_contents, part.file_key, part.file_bucket)
+                del file_contents
+
+                if '.epub' in part.file_key:
+                    file_root = '.'.join(part.file_key.split('.')[:-1])
+
+                    web_pub_manifest = self.generate_webpub(file_root)
+                    self.storage_manager.put_object(web_pub_manifest, f'{file_root}/manifest.json', self.file_bucket)
+
+            logger.info(f'Stored file at {part.url} from {part.source_url}')
         except Exception as e:
             logger.exception(f'Failed to store file {part.file_key} from {part.source_url}')
             raise e
+        
+    def _store_file_from_drive(self, part: Part):
+        file_id = f'{self.drive_service.id_from_url(part.source_url)}'
+        file = self.drive_service.get_drive_file(file_id)
+
+        if not file:
+            raise Exception(f'Unable to get file for: {part}')
+
+        file_permissions = None if part.file_bucket == self.limited_file_bucket else 'public-read'
+        self.storage_manager.put_object(file.getvalue(), part.file_key, part.file_bucket, bucket_permissions=file_permissions)
```

Comment:
just fyi, you can use `s3.upload_fileobj` to directly send the bytesio stream, instead of loading all the bytes and passing them to s3.  I *think* that also results in a multipart upload, but not sure tbh lol

Reply:
Good call - changed!